<!-- ===============================
Couverture
=============================== -->
<div style="background-color:#004080; color:white; padding:20px; border-radius:10px; text-align:center;">
  <h1>Audit linguistique du sous-ensemble Wolof</h1>
  <h3>Dataset MURI-IT — Reverse Instruction</h3>
  <p style="font-size:16px;">Auteur : Marème DIOP | Date : 21/02/2026</p>
  <p style="font-size:14px; font-style:italic;">
  Ce rapport présente une évaluation méthodologique du sous-ensemble étiqueté « Wolof »
  dans le dataset MURI-IT, afin d’examiner la cohérence entre l’étiquetage linguistique
  et le contenu effectif des données.
  </p>
</div>

---

<!-- ===============================
Contexte scientifique
=============================== -->
<h2 id="contexte" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Contexte scientifique
</h2>

<p>
Le dataset <strong>MURI-IT</strong> repose sur un mécanisme de génération d’instructions par
<em>Reverse Instruction</em>, appliqué à des corpus multilingues. Dans ce type de pipeline,
la validité linguistique des sous-ensembles générés dépend étroitement du système
de détection automatique de langue (Language Identification — LID) utilisé pour
l’étiquetage et la validation des données.
</p>

<p>
Dans le cas des langues à faibles ressources, telles que le wolof,
les performances des détecteurs généralistes peuvent varier,
et une validation insuffisamment robuste peut entraîner une discordance
entre l’étiquetage linguistique et le contenu réel des textes.
Une telle discordance affecte directement la qualité du corpus
et la fiabilité des étapes ultérieures d’entraînement ou d’évaluation.
</p>

<p>
Le présent travail s’inscrit dans une démarche d’audit méthodologique
visant à examiner la cohérence linguistique du sous-ensemble
étiqueté « Wolof » dans MURI-IT.
L’analyse porte conjointement sur les instructions générées
et sur les documents sources associés.
</p>

---

<!-- ===============================
Objectif et hypothèses
=============================== -->
<h2 id="objectif" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;"> Objectif et hypothèses
</h2>

<p>
L’objectif de cette étude est d’évaluer empiriquement
si le sous-ensemble étiqueté « Wolof »
correspond effectivement à un corpus majoritairement rédigé en wolof,
et d’examiner la robustesse de son étiquetage linguistique.
</p>

<p>
Afin de structurer l’analyse, les hypothèses suivantes sont retenues :
</p>

<ul>
<li>
<strong>H1 — Discordance linguistique:
  </strong>
Le sous-ensemble étiqueté « Wolof » présente une discordance significative
entre l’étiquetage linguistique et le contenu réel des textes.
</li>

<li>
<strong>H2 — Cohérence de validation :</strong>
L’étiquetage linguistique du sous-ensemble est cohérent
avec les prédictions de détecteurs de langue indépendants et robustes.
</li>

<li>
<strong>H3 — Détectabilité du signal :</strong>
Si le sous-ensemble est effectivement majoritairement rédigé en wolof,
alors un signal linguistique wolof devrait être détectable de manière robuste
par des classifieurs multilingues indépendants,
avec des niveaux de confiance substantiels.
</li>
</ul>

<p>
L’analyse qui suit mobilise plusieurs détecteurs de langue indépendants
afin de tester ces hypothèses de manière quantitative et comparative.</p>

<!-- ===============================
Cadre méthodologique
=============================== -->
<h2 id="methodologie" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Cadre méthodologique
</h2>

<p>
L’évaluation repose sur une stratégie hiérarchique de détection linguistique
visant à tester la robustesse du signal wolof dans un espace linguistique ouvert,
puis dans un espace spécialisé.
</p>

<p>
L’utilisation exclusive d’un détecteur spécialisé pour les langues africaines
n’est pas suffisante dans ce contexte. En effet, ces modèles opèrent dans un
espace restreint et ne couvrent pas l’ensemble des langues globalement dominantes,
notamment l’anglais et le français. Lorsqu’un texte appartient à une langue
non couverte par le modèle, celui-ci peut produire une attribution contrainte
vers la classe la moins improbable, sans que cela ne reflète un signal linguistique réel.
</p>

<p>
Afin d’éviter ce biais structurel, l’analyse procède en trois étapes :
</p>

<ul>
<li>
<strong>Étape 1 — Filtrage multilingue global :</strong>
Application d’un classifieur couvrant un large ensemble de langues,
afin d’identifier les langues dominantes dans un espace ouvert
et de détecter la présence éventuelle de langues non africaines.
</li>

<li>
<strong>Étape 2 — Calibration par seuil de confiance :</strong>
Séparation des prédictions robustes des cas ambigus,
sur la base des niveaux de confiance fournis par le classifieur global.
</li>

<li>
<strong>Étape 3 — Analyse spécialisée africaine :</strong>
Application d’un détecteur spécialisé sur les cas résiduels,
afin d’examiner la présence d’un signal wolof dans un espace linguistique restreint.
</li>
</ul>

<p>
Cette approche hiérarchique permet de distinguer les langues
clairement identifiables à l’échelle globale des cas nécessitant
une analyse intra-africaine plus fine,
et d’évaluer la détectabilité effective du signal wolof
dans le sous-ensemble étudié.
</p>

<!-- ===============================
Table des matières
=============================== -->
<div style="background-color:#f0f0f0; padding:15px; border-radius:8px; line-height:1.6;">
<h2 style="color:#004080;">📑 Table des matières</h2>

<ol>
<li><a href="#contexte">Contexte scientifique</a></li>

<li><a href="#objectif">Objectif et hypothèses</a></li>

<li><a href="#dataset">Présentation du sous-ensemble étudié</a></li>
<li><a href="#models">Modèles de détection de langue utilisés</a></li>

<li><a href="#fasttext_detection">Détection multilingue globale (fastText)</a></li>

<li><a href="#afrolid_analysis">Analyse spécialisée des cas ambigus (AfroLID)</a></li>

<li><a href="#triangulation">Analyse comparative et triangulation</a></li>

<li><a href="#discussion">Discussion méthodologique</a></li>

<li><a href="#conclusion">Conclusion générale</a></li>
</ol>

</div>

<!-- ===============================
Présentation du sous-ensemble étudié
=============================== -->
<h2 id="dataset" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Présentation du sous-ensemble étudié
</h2>

<p>
Le présent audit porte sur le sous-ensemble étiqueté <strong>« Wolof »</strong>
au sein du dataset <strong>MURI-IT</strong>, issu d’un pipeline de génération
par <em>Reverse Instruction</em>. Dans ce mécanisme, un document source est sélectionné,
traduit vers l’anglais, associé à une instruction générée automatiquement,
puis retraduit dans la langue cible.
</p>

<p>
Chaque exemple du sous-ensemble comprend deux composantes distinctes :
</p>

<ul>
<li><strong>Instruction (input)</strong> — requête générée automatiquement.</li>
<li><strong>Output (document source)</strong> — texte issu du corpus initial.</li>
</ul>

<p>
La distinction entre ces deux composantes est maintenue dans l’analyse,
afin d’examiner séparément les effets potentiels du mécanisme de génération
sur la validité linguistique.
</p>

---

<h3 style="color:#004080;">Taille et structure du sous-ensemble</h3>

<ul>
<li><strong>Nombre total d’exemples analysés :</strong> 4457</li>
<li><strong>Nombre d’instructions :</strong> 4457</li>
<li><strong>Nombre d’outputs :</strong> 4457</li>
</ul>

<p>
La correspondance un-à-un entre instructions et outputs est conservée
dans l’ensemble analysé.
</p>

---

<h3 style="color:#004080;">Caractéristiques textuelles globales</h3>

<p>
Les instructions présentent une longueur moyenne d’environ
<strong>144 caractères</strong> (≈ 25 mots), correspondant à des requêtes relativement brèves.
Les outputs présentent une longueur moyenne d’environ
<strong>481 caractères</strong>, avec une distribution plus hétérogène.
</p>

<p>
Ces caractéristiques descriptives fournissent un cadre contextuel
à l’analyse linguistique présentée dans les sections suivantes.
</p>

In [1]:
# ============================================================
# Chargement des bibliothèques
# ============================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datasets import load_dataset

pd.set_option("display.max_colwidth", 200)

print("Bibliothèques chargées avec succès.")

Bibliothèques chargées avec succès.


In [7]:
# ============================================================
# Chargement du dataset MURI-IT
# ============================================================

# Option 1 : Chargement depuis HuggingFace

df_wolof = pd.read_csv("/content/muri_wolof.csv")

print("Dataset chargé.")


Dataset chargé.


In [3]:


print(f"Nombre d'exemples étiquetés Wolof : {len(df_wolof)}")

Nombre d'exemples étiquetés Wolof : 4457


In [4]:
# ============================================================
# Séparation des instructions et des outputs
# ============================================================

instructions = df_wolof["input"].astype(str)
outputs = df_wolof["output"].astype(str)

print("Séparation effectuée.")
print(f"Nombre d'instructions : {len(instructions)}")
print(f"Nombre d'outputs : {len(outputs)}")

Séparation effectuée.
Nombre d'instructions : 4457
Nombre d'outputs : 4457


In [5]:
# ============================================================
# Statistiques descriptives préliminaires
# ============================================================

def compute_basic_stats(series, name):
    lengths_chars = series.apply(len)
    lengths_words = series.apply(lambda x: len(x.split()))

    print(f"\nStatistiques pour : {name}")
    print("-" * 40)
    print(f"Longueur moyenne (caractères) : {lengths_chars.mean():.2f}")
    print(f"Longueur médiane (caractères) : {lengths_chars.median():.2f}")
    print(f"Longueur moyenne (mots) : {lengths_words.mean():.2f}")
    print(f"Longueur médiane (mots) : {lengths_words.median():.2f}")

compute_basic_stats(instructions, "Instructions")
compute_basic_stats(outputs, "Outputs")


Statistiques pour : Instructions
----------------------------------------
Longueur moyenne (caractères) : 144.06
Longueur médiane (caractères) : 136.00
Longueur moyenne (mots) : 25.30
Longueur médiane (mots) : 24.00

Statistiques pour : Outputs
----------------------------------------
Longueur moyenne (caractères) : 481.49
Longueur médiane (caractères) : 105.00
Longueur moyenne (mots) : 92.38
Longueur médiane (mots) : 21.00


<!-- ===============================
Modèles de détection utilisés
=============================== -->
<h2 id="models" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Modèles de détection de langue utilisés
</h2>

<p>
L’analyse repose sur une stratégie hiérarchique mobilisant deux détecteurs complémentaires :
</p>

<ul>
<li><strong>fastText LID (176 langues)</strong> — utilisé en première étape pour une détection multilingue globale, incluant les langues à forte ressource (anglais, français, etc.).</li>
<li><strong>AfroLID</strong> — utilisé en seconde étape pour une analyse spécialisée des langues africaines, notamment le wolof.</li>
</ul>

<p>
fastText permet d’identifier les langues dominantes non couvertes par AfroLID,
tandis qu’AfroLID est mobilisé pour affiner l’analyse sur les cas résiduels.
Cette combinaison vise à limiter les biais liés à la couverture linguistique des modèles.
</p>

<p>
Le choix d’AfroLID s’appuie sur une évaluation comparative préalable des détecteurs de langue pour le wolof :
<br>
<a href="https://github.com/MaDi-12/wolof-language-detection-eval" target="_blank">
Benchmark des détecteurs de langue pour le Wolof
</a>
</p>

<p>
La section suivante présente les résultats de la détection multilingue globale.
</p>

<!-- ===============================
Détection multilingue globale
=============================== -->
<h2 id="fasttext_detection" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Détection multilingue globale
</h2>

<p>
Une première analyse est réalisée à l’aide du classifieur multilingue
<strong>fastText LID (176 langues)</strong>, afin d’identifier
la distribution globale des langues présentes dans le sous-ensemble
étiqueté « Wolof ».
</p>

<p>
Cette étape vise à tester l’existence d’un signal linguistique
dominant compatible avec l’étiquetage initial,
et à détecter d’éventuelles langues à forte ressource
susceptibles de contaminer le sous-ensemble.
</p>

---

<h3 style="color:#004080;">Application d’un seuil de confiance</h3>

<p>
Afin de distinguer les prédictions robustes des cas incertains,
un seuil de confiance <strong>τ = 0.7</strong> est appliqué
aux probabilités retournées par fastText.
</p>

<p>
Les instances dont la probabilité maximale est supérieure ou égale à τ
sont considérées comme des prédictions à haute confiance.
Les autres sont classées comme cas ambigus
et feront l’objet d’une analyse complémentaire.
</p>

---

<h3 style="color:#004080;">Résultats globaux</h3>

<p>
La distribution des langues détectées par fastText
est présentée ci-dessous.
</p>

In [6]:
!pip install transformers==4.37.2 --quiet

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.4/129.4 kB 11.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.4/8.4 MB 104.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.4/566.4 kB 49.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 132.4 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
sentence-transformers 5.2.3 requires transformers<6.0.0,>=4.41.0, but you have transformers 4.37.2 which is incompatible.


In [7]:
# ===============================
# Chargement du détecteur global (fastText LID 176)
# ===============================

!pip install numpy==1.26.4 --quiet
!pip install fasttext --quiet
!wget -q https://dl.fbaipublicfiles.com/fasttext/supervised-models/lid.176.bin



     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.0/18.0 MB 40.6 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
opencv-python-headless 4.13.0.92 requires numpy>=2; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.
xarray-einstats 0.10.0 requires numpy>=2.0, but you have numpy 1.26.4 which is incompatible.
pytensor 2.38.2 requires numpy>=2.0, but you have numpy 1.26.4 which is incompatible.
opencv-contrib-python 4.13.0.92 requires numpy>=2; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.
rasterio 1.5.0 requires numpy>=2, but you have numpy 1.26.4 which is incompatible.
opencv-python 4.13.0.92 requires numpy>=2; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.
cupy-cuda12x 14.0.1 requires numpy<2

In [5]:
import fasttext

ft_model = fasttext.load_model("lid.176.bin")


In [8]:
# ===============================
# Détection de langue avec fastText
# ===============================

from tqdm import tqdm
import pandas as pd

def detect_language_fasttext(text):
    try:
        labels, probabilities = ft_model.predict(str(text).replace("\n", " "), k=1)

        language = labels[0].replace("__label__", "")
        confidence = float(probabilities[0])

        return language, confidence

    except Exception as e:
        return "error", 0.0

In [9]:
from tqdm import tqdm
import pandas as pd

tqdm.pandas()

df_wolof[["ft_pred_lang", "ft_confidence"]] = (
    df_wolof["input"]
    .progress_apply(lambda x: pd.Series(detect_language_fasttext(x)))
)

100%|██████████| 4457/4457 [00:00<00:00, 8781.31it/s] 


In [10]:
# ===============================
# Résumé statistique — fastText (Instructions)
# ===============================

total = len(df_wolof)

# Distribution complète des langues
language_counts = df_wolof["ft_pred_lang"].value_counts()
language_percentages = df_wolof["ft_pred_lang"].value_counts(normalize=True) * 100

summary_df = pd.DataFrame({
    "Nombre": language_counts,
    "Pourcentage (%)": language_percentages.round(2)
})

print("📊 Distribution complète des langues détectées (fastText)")
display(summary_df.head(10))

# Taux de wolof
wolof_detected = language_counts.get("wo", 0)
wolof_rate = wolof_detected / total

print("\n" + "-"*50)
print("📌 Indicateurs clés")
print(f"Nombre total d’instructions analysées : {total}")
print(f"Nombre détecté comme wolof : {wolof_detected}")
print(f"Taux de wolof détecté : {wolof_rate:.2%}")
print(f"Taux estimé de non-wolof : {(1 - wolof_rate):.2%}")

# Analyse des scores de confiance
print("\n" + "-"*50)
print("📈 Statistiques des scores de confiance")
display(df_wolof["ft_confidence"].describe())

📊 Distribution complète des langues détectées (fastText)


,Nombre,Pourcentage (%)
ft_pred_lang,,
en,1924,43.17
fr,444,9.96
cy,140,3.14
it,107,2.40
tl,104,2.33
zh,93,2.09
sw,84,1.88
ur,84,1.88
es,74,1.66



--------------------------------------------------
📌 Indicateurs clés
Nombre total d’instructions analysées : 4457
Nombre détecté comme wolof : 0
Taux de wolof détecté : 0.00%
Taux estimé de non-wolof : 100.00%

--------------------------------------------------
📈 Statistiques des scores de confiance


,ft_confidence
count,4457.000000
mean,0.573474
std,0.333502
min,0.045478
25%,0.246158
50%,0.528536
75%,0.934624
max,1.000067


In [11]:
# ===============================
# Séparation par seuil de confiance (fastText)
# ===============================

CONFIDENCE_THRESHOLD = 0.7

# Marquage haute / basse confiance
df_wolof["ft_high_confidence"] = df_wolof["ft_confidence"] >= CONFIDENCE_THRESHOLD

# Séparation des sous-ensembles
df_high_conf = df_wolof[df_wolof["ft_high_confidence"]].copy()
df_low_conf = df_wolof[~df_wolof["ft_high_confidence"]].copy()

# Statistiques
total = len(df_wolof)
high_count = len(df_high_conf)
low_count = len(df_low_conf)

print("📊 Résumé de la séparation par seuil")
print("-" * 50)
print(f"Seuil de confiance τ = {CONFIDENCE_THRESHOLD}")
print(f"Total instructions : {total}")
print(f"Haute confiance (≥ τ) : {high_count} ({high_count/total:.2%})")
print(f"Cas ambigus (< τ) : {low_count} ({low_count/total:.2%})")

📊 Résumé de la séparation par seuil
--------------------------------------------------
Seuil de confiance τ = 0.7
Total instructions : 4457
Haute confiance (≥ τ) : 1912 (42.90%)
Cas ambigus (< τ) : 2545 (57.10%)


In [12]:
df_low_conf.sample(20)

,input,output,dataset_name,subdataset_name,language,split,language_name,ft_pred_lang,ft_confidence,ft_high_confidence
104,Naswona a ku laveki leswaku u kuma nomboro ya ...,Amul tamit dara lu laaj nga am nimaro ci nit y...,xP3,NaN,wol,train,Wolof,en,0.147527,False
1962,"Jaare ci saxaab. War a taxal ba ca xam-xam bi,...",Ñaqu rougeole (ŋas) mooy li lay musal ci rouge...,MRI,wikipedia,wol,train,Wolof,cy,0.147673,False
3958,"Loko a wu yime eka mahetelelo wa lwandle, a wu...","Kon sudee da nga taxawoon ci wetu ab deq, di n...",xP3,NaN,wol,train,Wolof,en,0.448305,False
1240,北京将举办开幕式、闭幕式和室内滑冰赛事。 Here is the same text in ...,"Pékin dina dalal feet ubbité bi ak tëjbi, dina...",xP3,NaN,wol,train,Wolof,zh,0.488075,False
748,War a paragraaf ci duniyaa gi ca karni 19.,"Sunu dund gi, niki gu dundat yi ak gu gàncax, ...",MRI,wikipedia,wol,train,Wolof,en,0.133934,False
963,Kugwiliskira ntchito mavidiyo ghakujambulika k...,Jëfandikoob dencug wideo buñuy aarësistere tax...,xP3,NaN,wol,train,Wolof,en,0.200332,False
320,"A text in Tumbuka: Wakayowoyaso kuti ""kweni vy...","Mu dellu waxni ""kenn waru leen woon wax ñu nan...",xP3,NaN,wol,train,Wolof,en,0.218547,False
252,"Fatou Diome, moom mi ngi leen di ko waxte ci 2...",Faatu Joom mi juddu ci atum 1968 ca Ñoojoor ci...,MRI,wikipedia,wol,train,Wolof,it,0.147550,False
4074,"Won na ko ci réew mi? Kon nga xam ne, ñiy waxt...",Concepción dëkku di Sili. Nitñii motnañu 261.0...,MRI,wikipedia,wol,train,Wolof,cy,0.123211,False
1681,"Text in French: Inouïe, La face cachée de la l...",Salacious li moon nudeli moonli moon wow tribb...,xP3,NaN,wol,train,Wolof,en,0.349327,False


<!-- ===============================
Interprétation — Détection multilingue globale
=============================== -->
<h3 style="color:#004080; border-bottom:2px solid #004080; padding-bottom:4px;" id="">
Interprétation des résultats — Détection globale (fastText)
</h3>

<p>
L’analyse multilingue réalisée sur les 4 457 instructions
montre une distribution linguistique dominée par des langues
à forte ressource.
</p>

<ul>
<li>Anglais : 43,17 % (1 924 instances)</li>
<li>Français : 9,96 % (444 instances)</li>
<li>Autres langues diverses : proportions inférieures à 4 % chacune</li>
</ul>

<p>
Aucune instance n’est détectée comme wolof par fastText,
soit un taux de wolof détecté de 0,00 %.
</p>

---

<h4 style="color:#004080;">Analyse des niveaux de confiance</h4>

<p>
Le score moyen de confiance est de 0,57 (écart-type = 0,33),
avec une médiane à 0,53.
Le troisième quartile atteint 0,93,
indiquant qu’une fraction substantielle des prédictions
présente une probabilité élevée.
</p>

<p>
Après application d’un seuil τ = 0,7 :
</p>

<ul>
<li>42,90 % des instructions (1 912) sont classées à haute confiance</li>
<li>57,10 % (2 545) constituent des cas ambigus</li>
</ul>

---

<h4 style="color:#004080;">Interprétation méthodologique</h4>

<p>
La dominance de l’anglais et du français,
associée à l’absence totale de détection du wolof,
indique qu’aucun signal wolof robuste
n’est identifiable par un classifieur multilingue global.
</p>

<p>
Les prédictions à haute confiance suggèrent que
près de la moitié du sous-ensemble est linguistiquement
cohérente avec des langues non-wolof.
</p>

<p>
La proportion majoritaire de cas ambigus (57,10 %)
justifie une analyse complémentaire,
afin de déterminer si un signal wolof
faiblement capté par fastText
pourrait être détecté par un modèle spécialisé.
</p>

<p>
Ces résultats fournissent un premier élément empirique
en faveur de l’hypothèse H1
(discordance entre étiquetage et contenu),
tout en nécessitant une vérification
dans le cadre de l’hypothèse H3.
</p>

<!-- ===============================
Analyse spécialisée des cas ambigus
=============================== -->
<h2 id="afrolid_analysis" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Analyse spécialisée des cas ambigus
</h2>

<p>
Les résultats obtenus avec fastText montrent qu’une proportion majoritaire
des instructions (57,10 %) présente un niveau de confiance inférieur au seuil
τ = 0,7.
</p>

<p>
Ces 2 545 instances constituent un sous-ensemble
linguistiquement incertain,
pour lequel la détection multilingue globale
ne permet pas de conclure de manière robuste.
</p>

<p>
Afin d’examiner l’existence éventuelle
d’un signal wolof faiblement capté par fastText,
nous appliquons AfroLID
à ces cas ambigus uniquement.
</p>

<p>
Cette étape vise à tester si un détecteur spécialisé
dans les langues africaines
identifie une proportion substantielle d’exemples
comme relevant du wolof.
</p>

<p>
Les résultats sont présentés ci-dessous.
</p>

In [13]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification


In [28]:
!pip uninstall numpy -y --quiet
!pip install numpy==1.26.4 --quiet

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
opencv-python-headless 4.13.0.92 requires numpy>=2; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.
xarray-einstats 0.10.0 requires numpy>=2.0, but you have numpy 1.26.4 which is incompatible.
pytensor 2.38.2 requires numpy>=2.0, but you have numpy 1.26.4 which is incompatible.
opencv-contrib-python 4.13.0.92 requires numpy>=2; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.
rasterio 1.5.0 requires numpy>=2, but you have numpy 1.26.4 which is incompatible.
opencv-python 4.13.0.92 requires numpy>=2; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.
cupy-cuda12x 14.0.1 requires numpy<2.6,>=2.0, but you have numpy 1.26.4 which is incompatible.
tobler 0.13.0 requires numpy>=2.0, but you have numpy 1.26.4 which is incompatible.
jax 0.7.2 requir

In [2]:
from transformers import pipeline
from tqdm import tqdm
tqdm.pandas()

classifier_afrolid = pipeline(
    "text-classification",
    model="UBC-NLP/afrolid_1.5",
    return_all_scores=True
)

print("AfroLID 1.5 chargé depuis HuggingFace.")

config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.11G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/514 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/4.80M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/18.0M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/279 [00:00<?, ?B/s]

AfroLID 1.5 chargé depuis HuggingFace.


/usr/local/lib/python3.12/dist-packages/transformers/pipelines/text_classification.py:104: UserWarning: `return_all_scores` is now deprecated,  if want a similar functionality use `top_k=None` instead of `return_all_scores=True` or `top_k=1` instead of `return_all_scores=False`.
  warnings.warn(


In [3]:
# ============================================================
# Fonction de prédiction robuste avec gestion des textes longs
# Agrégation : max score parmi les segments
# ============================================================

def split_text(text, max_chars=300):
    """
    Découpe un texte en segments de longueur max_chars.
    """
    return [text[i:i+max_chars] for i in range(0, len(text), max_chars)]


def predict_afrolid(texts, max_chars=300):
    predicted_languages = []
    confidence_scores = []

    for text in tqdm(texts):

        best_label = None
        best_score = 0

        if len(text) <= max_chars:

            results = classifier_afrolid(text)[0]

            best_entry = max(results, key=lambda x: x['score'])
            best_label = best_entry['label']
            best_score = best_entry['score']

        else:

            segments = split_text(text, max_chars=max_chars)

            for seg in segments:

                results = classifier_afrolid(seg)[0]

                seg_best = max(results, key=lambda x: x['score'])

                if seg_best['score'] > best_score:
                    best_score = seg_best['score']
                    best_label = seg_best['label']

        predicted_languages.append(best_label)
        confidence_scores.append(best_score)

    return predicted_languages, confidence_scores

In [14]:
# ===============================
# Vérification AfroLID sur cas ambigus et analyse des résultats
# ===============================

afrolid_predicted_langs, afrolid_confidence_scores = predict_afrolid(df_low_conf["input"].tolist())

df_low_conf["afrolid_lang"] = afrolid_predicted_langs
df_low_conf["afrolid_conf"] = afrolid_confidence_scores

print("✅ Vérification AfroLID des cas ambigus terminée.")

# ===============================
# Analyse des résultats AfroLID pour les cas ambigus
# ===============================

total_low_conf = len(df_low_conf)

print(f"\n📊 Résumé AfroLID pour les {total_low_conf} cas ambigus (fastText confidence < {CONFIDENCE_THRESHOLD})")
print("-" * 70)


lang_counts_afrolid = df_low_conf["afrolid_lang"].value_counts()
lang_percentages_afrolid = df_low_conf["afrolid_lang"].value_counts(normalize=True) * 100

summary_afrolid_df = pd.DataFrame({
    "Nombre": lang_counts_afrolid,
    "Pourcentage (%)": lang_percentages_afrolid.round(2)
})

print("Distribution des langues détectées par AfroLID :")
display(summary_afrolid_df.head(10))

# Taux de wolof détecté par AfroLID parmi les cas ambigus
# .get() will now work correctly as the index contains strings
wolof_detected_afrolid = lang_counts_afrolid.get("wol", 0)
wolof_rate_afrolid = wolof_detected_afrolid / total_low_conf

print("\n" + "-"*70)
print("📌 Indicateurs clés (AfroLID sur cas ambigus)")
print(f"Instructions détectées comme Wolof par AfroLID : {wolof_detected_afrolid}")
print(f"Taux de Wolof détecté par AfroLID : {wolof_rate_afrolid:.2%}")

# Analyse des scores de confiance d'AfroLID
print("\n" + "-"*70)
print("📈 Statistiques des scores de confiance AfroLID")
display(df_low_conf["afrolid_conf"].describe())

# Impact global estimé
total_global = len(df_wolof)
# We sum up the high confidence wolof detections from fastText (if any) and the low confidence wolof detections from AfroLID
# Assuming df_high_conf was already processed and we know its wolof count
# As per earlier FastText results, wolof_detected for fasttext was 0. So, we only consider wolof_detected_afrolid here.
wolof_global_estimated = wolof_detected_afrolid # Since fastText found 0 wolof at high confidence
wolof_global_rate = wolof_global_estimated / total_global

print("\n" + "-"*70)
print("🌍 Estimation globale après triangulation")
print(f"Wolof détecté (global) : {wolof_global_estimated}")
print(f"Taux global estimé de wolof : {wolof_global_rate:.2%}")

100%|██████████| 2545/2545 [06:43<00:00,  6.30it/s]

✅ Vérification AfroLID des cas ambigus terminée.

📊 Résumé AfroLID pour les 2545 cas ambigus (fastText confidence < 0.7)
----------------------------------------------------------------------
Distribution des langues détectées par AfroLID :


,Nombre,Pourcentage (%)
afrolid_lang,,
wol,1282,50.37
tsn,148,5.82
kin,123,4.83
twi,98,3.85
ibo,89,3.50
swh,69,2.71
zul,59,2.32
fon,59,2.32
lin,57,2.24



----------------------------------------------------------------------
📌 Indicateurs clés (AfroLID sur cas ambigus)
Instructions détectées comme Wolof par AfroLID : 1282
Taux de Wolof détecté par AfroLID : 50.37%

----------------------------------------------------------------------
📈 Statistiques des scores de confiance AfroLID


,afrolid_conf
count,2545.000000
mean,0.897019
std,0.205977
min,0.091407
25%,0.962951
50%,0.997039
75%,0.998649
max,0.999996



----------------------------------------------------------------------
🌍 Estimation globale après triangulation
Wolof détecté (global) : 1282
Taux global estimé de wolof : 28.76%


In [15]:
df_strong_wolof = df_low_conf[
    (df_low_conf["afrolid_lang"] == "wol") &
    (df_low_conf["afrolid_conf"] >= 0.5)
]

print("Nombre wolof fort :", len(df_strong_wolof))

Nombre wolof fort : 1120


In [16]:
# ===============================
# Distribution complète AfroLID (cas ambigus)
# ===============================

afro_counts_full = df_low_conf["afrolid_lang"].value_counts()
afro_percent_full = df_low_conf["afrolid_lang"].value_counts(normalize=True) * 100

afro_full_summary = pd.DataFrame({
    "Nombre": afro_counts_full,
    "Pourcentage (%)": afro_percent_full.round(2)
})

print("📊 Distribution complète des langues détectées par AfroLID")
display(afro_full_summary)

📊 Distribution complète des langues détectées par AfroLID


,Nombre,Pourcentage (%)
afrolid_lang,,
wol,1282,50.37
tsn,148,5.82
kin,123,4.83
twi,98,3.85
ibo,89,3.50
swh,69,2.71
zul,59,2.32
fon,59,2.32
lin,57,2.24


In [17]:
# ===============================
# Extraction des textes prédits "wol" par AfroLID
# ===============================

df_afro_wol = df_low_conf[df_low_conf["afrolid_lang"] == "wol"].copy()

print("Nombre total prédits 'wol' par AfroLID :", len(df_afro_wol))

cols_to_display = [
    "input",
    "ft_pred_lang",
    "ft_confidence",
    "afrolid_lang",
    "afrolid_conf"
]

display(df_afro_wol[cols_to_display].head(20))

Nombre total prédits 'wol' par AfroLID : 1282


,input,ft_pred_lang,ft_confidence,afrolid_lang,afrolid_conf
1,"दुर्घटनापछि, गिब्सनलाई अस्पतालमा लगियो तर केही...",ne,0.626493,wol,0.369746
4,Descrive Cova Figueira. Di ko war a faraaw ci ...,ro,0.151771,wol,0.995326
7,Ci lu tax a ngi leen diw ak liis bu baax ci de...,si,0.139856,wol,0.998298
8,Lu tax lu baax ci ay xaraj yi? Li mu ko diiwal...,eo,0.159266,wol,0.997685
9,"Defeend taariki Senegal. Défaat ci àll bi, ak ...",bn,0.085498,wol,0.994408
14,Ni ko leen di gën a wax? Won la ci niokolo-Kob...,it,0.152316,wol,0.998054
21,"Jaare ci kaw giir yi, te xam-xam na ko gën a b...",en,0.159976,wol,0.996315
22,"Eyi nti, ↄpɛpↄn da a edi kan na wobenya tumi k...",en,0.174856,wol,0.431044
25,Lu tax lu baax ci réew mi? Li mu ngi leen di k...,ia,0.253249,wol,0.998342
30,"Lu tax woon ci réew mi? Li ko gën a waxte, li ...",tr,0.350203,wol,0.997223


In [18]:
df_afro_wol.to_csv("afrolid_predicted_wolof_low_conf.csv", index=False)

<!-- ===============================
Interprétation — Analyse spécialisée AfroLID
=============================== -->
<h3 style="color:#004080; border-bottom:2px solid #004080; padding-bottom:4px;">
Interprétation des résultats — Analyse AfroLID
</h3>

<p>
L’application d’AfroLID aux  2 545 instances
classées comme ambiguës par fastText révèle
une structure linguistique fortement dominée
par des langues africaines, avec une proportion
majoritaire de wolof.
</p>

<p>
Parmi ces instances :
</p>

<ul>
<li>1 282 sont classées comme <strong>wol</strong> (50,37 %)</li>
<li>Les autres langues détectées incluent tsn (5,82 %), kin (4,83 %),
twi (3,85 %), ibo (3,50 %), ainsi qu’un ensemble de langues africaines
en proportions décroissantes.</li>
</ul>

---

<h4 style="color:#004080;">Analyse des niveaux de confiance</h4>

<p>
Contrairement à l’analyse précédente, les scores de confiance
associés aux prédictions AfroLID sont ici élevés :
</p>

<ul>
<li>Moyenne : 0,897</li>
<li>Médiane : 0,997</li>
<li>3e quartile : 0,9986</li>
<li>Maximum : 0,9999</li>
</ul>

<p>
Ces valeurs indiquent que les prédictions sont fortement
confiantes et reposent sur un signal linguistique robuste.
La majorité des prédictions wolof sont associées
à des probabilités proches de 1, ce qui suggère
une détection fiable.
</p>

---

<h4 style="color:#004080;">Interprétation méthodologique</h4>

<p>
La divergence observée entre fastText et AfroLID
s’explique par la spécialisation du modèle AfroLID
sur les langues africaines, lui permettant de capter
un signal linguistique non détecté par fastText
dans les cas de faible confiance.
</p>

<p>
Ainsi, contrairement à l’hypothèse initiale d’absence
de signal wolof, les résultats indiquent que :
</p>

<ul>
<li>une proportion significative des cas ambigus contient effectivement du wolof</li>
<li>ce signal est détecté avec un haut niveau de confiance par AfroLID</li>
</ul>

<p>
Après triangulation, la proportion globale
d’exemples détectés comme wolof s’élève à
<strong>28,76 % (1 282 / 4 457)</strong>,
dont <strong>1 120</strong> cas avec un niveau de confiance élevé.
</p>

<p>
Ces résultats nuancent fortement l’hypothèse H1
(d’absence de contenu wolof) et suggèrent plutôt
une limitation des détecteurs généralistes
dans la détection de langues africaines
dans des contextes ambigus.
</p>

<!-- ===============================
Analyse comparative et triangulation
=============================== -->
<h2 id="triangulation" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
 Analyse comparative et triangulation
</h2>

<p>
Les résultats issus des différentes étapes de détection
permettent d’évaluer la cohérence linguistique
du sous-ensemble « Wolof » dans une perspective d’audit.
</p>

---

<h3 style="color:#004080;">Synthèse des observations</h3>

<ul>
<li>
Le filtrage multilingue global met en évidence
une présence importante de langues à forte ressource,
notamment l’anglais et le français,
avec une absence de détection significative du wolof
dans les prédictions à haute confiance.
</li>

<li>
L’analyse des cas ambigus via un détecteur spécialisé
révèle toutefois la présence d’un signal wolof,
représentant environ 50 % de ce sous-ensemble spécifique.
</li>

<li>
Après intégration des différentes étapes,
la proportion globale estimée de contenu wolof
s’élève à environ <strong>28,76 %</strong> du corpus analysé.
</li>
</ul>

---

<h3 style="color:#004080;">Interprétation du signal linguistique</h3>

<p>
Ces résultats indiquent que le wolof est effectivement présent
dans le sous-ensemble étudié,
mais qu’il ne constitue pas la langue majoritaire
à l’échelle globale du corpus.
</p>

<p>
Le signal wolof apparaît principalement
dans les cas initialement classés comme ambigus,
ce qui suggère qu’il est partiellement masqué
dans une analyse multilingue globale.
</p>

---

<h3 style="color:#004080;">Évaluation des hypothèses</h3>

<ul>
<li>
<strong>H1 — Discordance linguistique :</strong>
Confirmée partiellement.  
Une discordance significative est observée
entre l’étiquetage « Wolof » et la distribution réelle,
dans la mesure où le wolof ne constitue pas
la langue majoritaire du sous-ensemble.
</li>

<li>
<strong>H2 — Cohérence de validation :</strong>
Partiellement infirmée.  
L’étiquetage ne correspond pas pleinement
aux résultats obtenus via des détecteurs indépendants,
bien qu’un signal wolof réel soit présent.
</li>

<li>
<strong>H3 — Détectabilité du signal :</strong>
Confirmée.  
Un signal wolof est détectable de manière robuste,
mais uniquement dans une analyse adaptée
et non dans le filtrage global initial.
</li>
</ul>

---

<p>
Dans l’ensemble, la triangulation met en évidence
une présence réelle mais minoritaire du wolof,
suggérant une discordance quantitative
plutôt qu’une absence de signal linguistique.
</p>

<!-- ===============================
Discussion méthodologique
=============================== -->
<h2 id="discussion" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Discussion méthodologique
</h2>


<h3 style="color:#004080;">Lecture du désalignement observé</h3>

<p>
Le principal résultat de cet audit
réside dans l’écart entre l’étiquetage initial
du sous-ensemble « Wolof »
et la distribution linguistique observée.
</p>

<p>
La proportion globale de wolof (~28,76 %)
reste significative,
mais insuffisante pour caractériser le corpus
comme majoritairement wolof.
</p>

---

<h3 style="color:#004080;">Rôle du pipeline de génération</h3>

<p>
Le pipeline de Reverse Instruction
peut introduire des transformations linguistiques,
notamment via des langues pivot,
qui contribuent à la diversification linguistique
du corpus généré.
</p>

<p>
Ce phénomène peut expliquer
la présence simultanée de plusieurs langues,
y compris dans un sous-ensemble
étiqueté de manière unitaire.
</p>

---

<h3 style="color:#004080;">Portée de l’audit</h3>

<p>
Le présent travail ne remet pas en cause
l’existence de contenu en wolof,
mais met en évidence une discordance
dans sa proportion relative.
</p>

<p>
Il souligne la nécessité
d’une validation linguistique quantitative
lors de la construction de sous-ensembles
multilingues,
en particulier pour les langues
à faibles ressources.
</p>

<!-- ===============================
Conclusion générale
=============================== -->
<h2 id="conclusion" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Conclusion générale
</h2>

<p>
Ce travail avait pour objectif
d’évaluer la cohérence linguistique
du sous-ensemble étiqueté « Wolof »
dans le dataset MURI-IT.
</p>

<p>
L’analyse met en évidence
une distribution linguistique hétérogène,
dans laquelle le wolof est présent
mais ne constitue pas la langue dominante.
</p>

<p>
La proportion globale estimée (~28,76 %)
indique une discordance significative
entre l’étiquetage et le contenu réel du corpus.
</p>

<p>
Ces résultats confirment partiellement
l’hypothèse de discordance linguistique (H1),
invalident une cohérence stricte de validation (H2),
tout en confirmant la détectabilité du signal wolof (H3).
</p>

<p>
Ainsi, le sous-ensemble étudié
ne peut être considéré comme
majoritairement wolof,
mais plutôt comme un corpus mixte
incluant une composante wolof substantielle.
</p>

<p>
Ce constat souligne l’importance
de procédures de validation linguistique rigoureuses
dans la construction de datasets multilingues,
afin d’assurer leur fiabilité
pour les usages en apprentissage automatique.
</p>